src/en/news/blog/2025/rgw-improving-bucket-notifs/index.md

```diff
@@ -0,0 +1,218 @@
+---
+title: "Improving Persistent Bucket Notifications Parallelism"
+
+date: 2025-08-20
+author: Adarsh Ashokan
+tags:
+  - ceph
+  - rgw
+  - notifications
+---
+
+## Improving Persistent Bucket Notifications Parallelism
+
+### Problem at Hand  
+
+Persistent Bucket Notifications were first rolled out in **Ceph Pacific** as part of the **RADOS Gateway (RGW) project**.  
+For a good background, see the [existing blog post on ceph.io](https://ceph.io/en/news/blog/2021/persistent-bucket-notifications-deep-dive/).  
+
+In brief, this post deals with **improving the performance** of persistent bucket notifications.  
+
+Currently, the implementation doesn’t fully leverage RADOS’ distributed capabilities. Each topic is tied to a **2 Phase Commit Queue** implemented as a   **single RADOS object**.
+
+This design creates a bottleneck. A **sharded queue implementation** allows notifications to be distributed across multiple RADOS objects, enabling **parallel writes via multiple OSDs**.  
+
+---
+
+<!-- ### Example  
+
+- Suppose you create **10,000 objects** in a Ceph bucket.  
+- In the current design, all `"Object Created"` notifications are directed to a **single queue object**, causing contention on a single OSD.  
+- With **sharded topic queues**, notifications are split across multiple queue objects. Multiple OSDs now handle writes in parallel, removing the bottleneck.  
+
+--- -->
+
+### What Was Done  
+
+- In the **old design**, each topic was mapped to **one RADOS object** (one `2pc_cls_queue`).  
+- In the **new design**, each topic maps to **multiple RADOS objects** (multiple `2pc_cls_queue` objects).  
+
+Each `2pc_cls_queue` object associated with a topic is called a **shard**.  
+
+The number of shards is configurable via:  
+
+```bash
+rgw_bucket_persistent_notif_num_shards
+```
+
+- **Default value:** 11  
+- **Note:** Existing topics are **not** re-sharded. To extend this feature for existing topics, you must delete and recreate the topic.  
+
+---
+
+### Topic Management  
+
+Topic operations updated to support sharded queues:  
+
+| Operation        | Behavior                                                                    |
+| ---------------- | --------------------------------------------------------------------------- |
+| **Create topic** | Creates multiple `2pc_cls_queue` objects as shards                          |
+| **Delete topic** | Deletes all associated shards                                               |
+| **Set topic**    | Supports toggling persistent ↔ non-persistent, with shard cleanup if needed |
+
+---
+
+### Design Details  
+
+#### Enqueue  
+- Ordering is guaranteed at the **per-key level** (per object in a bucket).  
+- The **target shard** is computed as:  
```

Comment:
I'd probably do this:
s/The target shard is computed as/The target shard is computed using the following formula:/

I just think "as" runs the risk of being too vague.